pkg/kubelet/cm/devicemanager/topology_hints.go

```diff
@@ -81,10 +81,53 @@ func (m *ManagerImpl) GetTopologyHints(pod *v1.Pod, container *v1.Container) map
 	return deviceHints
 }
 
-// GetPodTopologyHints implements the TopologyManager HintProvider Interface which
+// GetPodTopologyHints implements the topologymanager.HintProvider Interface which
 // ensures the Device Manager is consulted when Topology Aware Hints for Pod are created.
 func (m *ManagerImpl) GetPodTopologyHints(pod *v1.Pod) map[string][]topologymanager.TopologyHint {
-	return nil
+	// Garbage collect any stranded device resources before providing TopologyHints
+	m.UpdateAllocatedDevices()
+
+	deviceHints := make(map[string][]topologymanager.TopologyHint)
+	accumulatedResourceRequests := m.getPodDeviceRequest(pod)
+
+	for resource, requested := range accumulatedResourceRequests {
+		// Only consider devices that actually contain topology information.
+		if aligned := m.deviceHasTopologyAlignment(resource); !aligned {
+			klog.Infof("[devicemanager] Resource '%v' does not have a topology preference", resource)
+			deviceHints[resource] = nil
+			continue
+		}
+
+		// Short circuit to regenerate the same hints if there are already
+		// devices allocated to the Pod. This might happen after a
+		// kubelet restart, for example.
+		allocated := m.podDevices.podDevices(string(pod.UID), resource)
+		if allocated.Len() > 0 {
+			if allocated.Len() != requested {
+				klog.Errorf("[devicemanager] Resource '%v' already allocated to (pod %v) with different number than request: requested: %d, allocated: %d", resource, format.Pod(pod), requested, allocated.Len())
+				deviceHints[resource] = []topologymanager.TopologyHint{}
+				continue
+			}
+			klog.Infof("[devicemanager] Regenerating TopologyHints for resource '%v' already allocated to (pod %v)", resource, format.Pod(pod))
+			deviceHints[resource] = m.generateDeviceTopologyHints(resource, allocated, sets.String{}, requested)
+			continue
+		}
+
+		// Get the list of available devices, for which TopologyHints should be generated.
+		available := m.getAvailableDevices(resource)
+		reusable := m.devicesToReuse[string(pod.UID)][resource]
+		if available.Len() < requested {
```

Comment:
Is there a reason that this line differs from the equivalent line in `GetTopologyHints()`?, i.e.:
```
if available.Union(reusable).Len() < requested {
```

I guess the notion of `reusable`  devices doesn't really make sense when we are considering things at the pod scope.

As such, this code block might be better implemented as:
```
        // Get the list of available devices, for which TopologyHints should be generated.
        available := m.getAvailableDevices(resource)
        if available.Len() < requested {
            klog.Errorf("[devicemanager] Unable to generate topology hints: requested number of devices unavailable for '%s': requested: %d, available: %d", resource, requested, availab
            deviceHints[resource] = []topologymanager.TopologyHint{}
            continue
        }

        // Generate TopologyHints for this resource given the current
        // request size and the list of available devices.
        deviceHints[resource] = m.generateDeviceTopologyHints(resource, available, sets.String{}, requested)
```

Reply:
I have applied the change, after quick thought, your point makes sense. I will check it more briefly though, to be sure.